tensorflow_datasets/scripts/metadata_cleanup.py

```diff
@@ -0,0 +1,97 @@
+# Lint as: python3
+"""Removes all dataset versions from metadata directory which are not 
+   present in registered versions of TFDS.
+
+To test:
+Goto tensorflow_datasets/scripts
+
+Args : dry_run : If given, it traverse and finds all
+                 versions which are to be removed without
+                 actually removing them.
+
+             r : If given it removes all unmatched dataset versions 
+                 from metadata directory.
+```
+cd tensorflow_datasets/scripts
+
+For check not delete : python metadata_cleanup.py --dry_run
+For delete not check : python metadata_cleaup.py --r
+For both delete & check : python metadata_cleanup.py --dry_run --r
+```
+"""
+from absl import flags
+from absl import app
+import os
+import tensorflow as tf
+from tensorflow_datasets.core import registered
+
+
+FLAGS = flags.FLAGS
+flags.DEFINE_boolean('dry_run', False, "Shows all versions\
+                      which are not registered")
+flags.DEFINE_boolean('r', False, "Removes unmatched\
+                      dataset versions from metadata")
+
+__metadata_path = os.path.join(os.pardir, "testing/metadata")
+
+def _extract_metadata_versions(metadata_dir):
+    '''
+       Get all metadata direcotry versions paths,
+       It only extract the paths like 'dataset_name/version'
+       or 'dataset_name/config/versions' in metadata dir
+    '''
+    _meta_paths = set()
+    for root, dirs, files in tf.io.gfile.walk(metadata_dir):
+        for fileName in files:

```

Comment:
Why iterating over `for fileName in files:` when you're not using it ?

Reply:
this is because to take only paths in format dataset/version because of `for fileName in files`  it traverse on each files and if we are not using them we are also getting directory path with `dataset/version` paths

Reply:
You should not have to do this if you're using `_validate_name` suggested bellow.
Also this is inefficient. `if len(files)` would be much better

Reply:
Matching patterns like I was done in my previous commit [here](https://github.com/tensorflow/datasets/pull/1730/commits/651711349336790414ece3b41c8dc38510abf739#diff-d3640dd24433be763b6ce93db70001c5L73-L80)